main/lsp/requests/completion.cc

```diff
@@ -92,36 +94,103 @@ string methodSnippet(const core::GlobalState &gs, core::SymbolRef method) {
-        if (prevNewlineLoc == preDefinition.npos) {
+    std::vector<string_view> documentation_lines;
+
+    // Iterate from the last line, to the first line
+    for (auto it = all_lines.rbegin(); it != all_lines.rend(); it++) {
+        string_view line = absl::StripAsciiWhitespace(*it);
+
+        // Short circuit when line is empty
+        if (line.empty()) {
+            break;
+        }
+
+        // Handle single-line sig block
+        else if (absl::StartsWith(line, "sig")) {
+            // Do nothing for a one-line sig block
+        }
+
+        // Handle multi-line sig block
+        else if (absl::StartsWith(line, "end")) {
+            // ASSUMPTION: We either hit the start of file, a `sig do` or an `end`
+            it++;
+            while (
+                // SOF
+                it != all_lines.rend()
+                // Start of sig block
+                && !absl::StartsWith(absl::StripAsciiWhitespace(*it), "sig do")
+                // Invalid end keyword
+                && !absl::StartsWith(absl::StripAsciiWhitespace(*it), "end")) {
+                it++;
+            };
+
+            // We have either
+            // 1) Reached the start of the file
+            // 2) Found a `sig do`
+            // 3) Found an invalid end keyword
+            if (it == all_lines.rend() || absl::StartsWith(absl::StripAsciiWhitespace(*it), "end")) {
+                break;
+            }
+
+            // Reached a sig block.
+            line = absl::StripAsciiWhitespace(*it);
+            ENFORCE(absl::StartsWith(line, "sig do"));
+
+            // Stop looking if this is a single-line block e.g `sig do; <block>; end`
+            if (absl::StartsWith(line, "sig do;") && absl::EndsWith(line, "end")) {
+                break;
+            }
+
+            // Else, this is a valid sig block. Move on to any possible documentation.
+        }
+
+        // Handle a comment line. Do not count typing declarations.
+        else if (absl::StartsWith(line, "#") && !absl::StartsWith(line, "# typed:")) {
+            // Account for whitespace before comment e.g
+            // # abc -> "abc"
+            // #abc -> "abc"
+            int skip_after_hash = absl::StartsWith(line, "# ") ? 2 : 1;
+
+            documentation_lines.push_back(line.substr(line.find('#') + skip_after_hash));
+        }
+
+        // No other cases applied to this line, so stop looking.
+        else {
             break;
         }
-        prevLine = preDefinition.substr(prevNewlineLoc, lastNewlineLoc - prevNewlineLoc);
-        lastNewlineLoc = prevNewlineLoc;
     }
-    if (documentation.empty()) {
+
+    string documentation = absl::StrJoin(documentation_lines.rbegin(), documentation_lines.rend(), "\n");
+    documentation = absl::StripTrailingAsciiWhitespace(documentation);
+
+    if (documentation.empty())
         return nullopt;
+    else {
+        // Ensure single newline for non-empty documentation
+        documentation.append("\n");
```

Comment:
I found a bug. If you have the following:

```
# This is a comment with a bulleted list:
# * Item 1
# * Item 2
# * Item 3
sig {void}
def my_method; end;
```

...then the markdown looks like:


This is a comment with a bulleted list:
* Item 1
* Item 2
* Item 3
  ```ruby
  sig {void}
  ```

(the sig is indented with the list)

I think we need two newlines between comments and the sig text.

Reply:
It's pretty weird, since:
```
* item 1
sentence after item 1
```

...gets rendered as:

* item 1 sentence after item 1

Reply:
I think this is a result of Markdown -- I think plaintext users won't complain too much if we add an extra newline :P

Reply:
Great catch
![image](https://user-images.githubusercontent.com/51730104/63068469-cd541880-bf45-11e9-960b-e6eea6f4aa02.png)

Reply:
TS has this cool interface, wonder if it's a bespoke thing or if they just use a markdown breakline 
![image](https://user-images.githubusercontent.com/51730104/63068563-32a80980-bf46-11e9-85f8-7002f651f663.png)